snapcraft/commands/lifecycle.py

```diff
@@ -0,0 +1,103 @@
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Snapcraft lifecycle commands."""
+
+import abc
+import textwrap
+from typing import TYPE_CHECKING
+
+from craft_cli import BaseCommand, emit
+from overrides import overrides
+
+from snapcraft_legacy.cli import legacy
+
+if TYPE_CHECKING:
+    import argparse
+
+
+class _LifecycleCommand(BaseCommand, abc.ABC):
+    """Run lifecycle commands."""
+
+    @overrides
+    def fill_parser(self, parser: "argparse.ArgumentParser") -> None:
+        parser.add_argument(
+            "parts", metavar="parts", type=str, nargs="*", help="Parts to process"
+        )
+
+    @overrides
+    def run(self, parsed_args):
+        """Run the command."""
+        if not self.name:
+            raise RuntimeError("command name not specified")
+
+        emit.trace(f"lifecycle command: {self.name!r}, arguments: {parsed_args!r}")
+        legacy.legacy_run()
+
+
+class PullCommand(_LifecycleCommand):
+    """Run the lifecycle up to the pull step."""
+
+    name = "pull"
+    help_msg = "Download or retrieve artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Download or retrieve artifacts defined for a part. If part names
+        are specified only those parts will be pulled, otherwise all parts
+        will be pulled.
+        """
+    )
+
+
+class BuildCommand(_LifecycleCommand):
+    """Run the lifecycle up to the build step."""
+
+    name = "build"
+    help_msg = "Build artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Build artifacts defined for a part. If part names are specified only
+        those parts will be built, otherwise all parts will be built.
+        """
+    )
+
+
+class StageCommand(_LifecycleCommand):
+    """Run the lifecycle up to the stage step."""
+
+    name = "stage"
+    help_msg = "Stage built artifacts into a common staging area"
+    overview = textwrap.dedent(
+        """
+        Stage built artifacts into a common staging area. If part names are
+        specified only those parts will be staged. The default is to stage
+        all parts.
+        """
+    )
+
+
+class PrimeCommand(_LifecycleCommand):
+    """Prepare the final payload for packing."""
+
+    name = "prime"
+    help_msg = "Build artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Prepare the final payload to be packed as a snap. If part names are
+        specified only those parts will be primed. The default is to prime
+        all parts.
```

Comment:
missing comments about also adding the metadata (snap.yaml, etc...)

Reply:
The metadata is to be added post-lifecycle, so the primed artifacts will not be the final payload to be packed as a snap. Will change the docstring.